fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,541 @@
+        cuda_stream = torch.cuda.current_stream(input.device) if input.device.type == "cuda" else None
+
+        with self.ready_cv:
+            assert self.tensors[i][j] is None
+            input, phony = fork(input)
+            self.batch_events[i][j] = cuda_stream.record_event() if cuda_stream is not None else None  # type: ignore
+            self.tensors[i][j] = input
+            self.ready_cv.notify_all()
+        return phony
+
+    def wait_for(self, chunk: int) -> None:
+        with self.ready_cv:
+            while self.batches[chunk] is None and any(b is None for b in self.tensors[chunk]):
+                self.ready_cv.wait()
+            if self.batches[chunk] is None:
+                tensors = cast(List[Tensor], self.tensors[chunk])
+                self.batches[chunk] = Batch(tuple(tensors), chunk)
+
+    def get_batch(self, chunk: int) -> Batch:
+        batch = self.batches[chunk]
+        assert batch is not None
+        return batch
+
+
+class PartitionHandler:
+    def __init__(
+        self,
+        module_rref: rpc.RRef,
+        device: str,
+        num_input: int,
+        num_output: int,
+        rank: int,
+        chunks: int,
+        checkpoint_stop: int,
+        loss_module_rref: Optional[rpc.RRef] = None,
+    ) -> None:
+        self.module = module_rref.local_value()
+        self.chunks = chunks
+        self.device = torch.device(device)
+        self.checkpoint_stop = checkpoint_stop
+        self.rank = rank
+        self.num_input = num_input
+        self.num_output = num_output
+        (self.in_queue,), (self.out_queue,) = create_workers([self.device])
+        self.loss_module = None if loss_module_rref is None else loss_module_rref.local_value()
+
+    def local_parameter_rrefs(self) -> List[rpc.RRef]:
+        r"""
+        Create one RRef for each parameter in the given local module, and return a
+        list of RRefs.
+        """
+        return [rpc.RRef(p) for p in self.module.parameters()]
+
+    def make_dist_record(self, users: List[Tuple[rpc.RRef, int, int]]) -> DistributedPipelineRecord:
+        return DistributedPipelineRecord(self.device, self.rank, self.chunks, self.num_input, users)
+
+    def run(self, dist_record: DistributedPipelineRecord) -> None:
+        """Runs pipeline parallelism.
+
+        It modifies the given batches in place.
+
+        """
+
+        m = len(dist_record.batches)
+
+        self.stream = current_stream(self.device)
+
+        for i in range(m):
+            with record_function("feed"):
+                dist_record.wait_for(i)
+            self.fence(dist_record, i)
+            self.compute(dist_record, i)
+            self.forward_results(i, dist_record)
+
+    def fence(self, dist_record: DistributedPipelineRecord, chunk: int) -> None:
+        """Copies micro-batches after computation for the previous
+        micro-batches.
+        """
+        # Ensure that batches[chunk-1] is executed after batches[chunk] in
```

Comment:
Why is this necessary?

Reply:
To prevent queuing in backward path. This makes sure microbatch no. chunk-1 is not copied over to this device in backward path before gradient of microbatch no. chunk is computed.